action/jsonapi.go

```diff
@@ -0,0 +1,210 @@
+	"encoding/json"
+	"fmt"
+	"github.com/justwatchcom/gopass/store"
+	"github.com/urfave/cli"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type messageType struct {
+	Type string `json:"type"`
+}
+
+type queryMessage struct {
+	Query string `json:"query"`
+}
+
+type getLoginMessage struct {
+	Entry string `json:"entry"`
+}
+
+type loginResponse struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+type errorResponse struct {
+	Error string `json:"error"`
+}
+
+// JSONAPI reads a json message on stdin and responds on stdout
+func (s *Action) JSONAPI(ctx context.Context, c *cli.Context) error {
+	err := readAndRespond(ctx, s)
+	if err != nil {
+		var response errorResponse
+		response.Error = err.Error()
+
+		err = sendSerializedJSONMessage(response)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func readAndRespond(ctx context.Context, s *Action) error {
+	message, err := readMessage()
+	if message != nil {
+		err = respondMessage(ctx, s, message)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func readMessage() ([]byte, error) {
+	stdin := bufio.NewReader(os.Stdin)
+	lenBytes := make([]byte, 4)
+	_, err := stdin.Read(lenBytes)
+	if err != nil {
+		return nil, eofReturn(err)
+	}
+	length, err := getMessageLength(lenBytes)
+	if err != nil {
+		return nil, err
+	}
+	msgBytes := make([]byte, length)
+	_, err = stdin.Read(msgBytes)
+	if err != nil {
+		return nil, eofReturn(err)
+	}
+	return msgBytes, nil
+}
+
+func getMessageLength(msg []byte) (int, error) {
+	var length uint32
+	buf := bytes.NewBuffer(msg)
+	err := binary.Read(buf, binary.LittleEndian, &length)
+	if err != nil {
+		return 0, err
+	}
+	return int(length), nil
+}
+
+func eofReturn(err error) error {
+	if err == io.EOF {
+		return nil
+	}
+	return err
+}
+
+func respondMessage(ctx context.Context, s *Action, msgBytes []byte) error {
+	var message messageType
+	err := json.Unmarshal(msgBytes, &message)
+	if err != nil {
+		return err
+	}
+	switch message.Type {
+	case "query":
+		return respondQuery(s, msgBytes)
+	case "getLogin":
+		return respondGetLogin(ctx, s, msgBytes)
+	default:
+		return fmt.Errorf("Unknown message of type %s", message.Type)
+	}
+}
+
+func respondQuery(s *Action, msgBytes []byte) error {
+	var message queryMessage
+	err := json.Unmarshal(msgBytes, &message)
+	if err != nil {
+		return err
+	}
+
+	l, err := s.Store.List(0)
+	if err != nil {
+		return err
+	}
+
+	needle := strings.ToLower(message.Query)
+	choices := make([]string, 0, 10)
+	for _, value := range l {
+		if strings.Contains(strings.ToLower(value), needle) {
+			choices = append(choices, value)
+		}
+	}
+
+	return sendSerializedJSONMessage(choices)
+}
+
+func respondGetLogin(ctx context.Context, s *Action, msgBytes []byte) error {
+	var message getLoginMessage
+	var response loginResponse
+
+	err := json.Unmarshal(msgBytes, &message)
+	if err != nil {
+		return err
+	}
+
+	secret, err := s.Store.Get(ctx, message.Entry)
+
```

Comment:
please remove this blank line, the error handling should visually stick to it's (possible) cause